### PR TITLE
Remove interpolated potentials

### PR DESCRIPTION
**Pull request by mcwitt**
- Removes `NonbondedInterpolated`, interpolated variants, and associated tests
   Note: this still leaves the computation of 4-d coordinates up to the GPU potentials. Plan is to update the signatures to accept `w_coords` and fully remove `k_lambda_transformer` in a future PR.
- Removes (previously deprecated) `InterpolatedPotential`

### CODE REVIEW

File: timemachine/lib/potentials.py [hunk 1]
```diff
@@ -63,81 +63,6 @@ def unbound_impl(self, precision):
...
-#     def get_N(self):
-#         return self.args[0]
-
-#     def get_a_idxs(self):
-#         return self.args[1]
-
-#     def get_b_idxs(self):
-#         return self.args[2]
-
-
```
Comment (mcwitt) [inline timemachine/lib/potentials.py:140]: Should `InterpolatedPotential` be moved to `attic/` (along with the custom ops implementation) or just deleted?

Comment (proteneer) [reply timemachine/lib/potentials.py:140]: Just delete it. I can't see a future use for this at all.

File: timemachine/cpp/src/math_utils.cuh [hunk 2]
```diff
@@ -1,13 +1,3 @@
 #pragma once
 
-#include "surreal.cuh"
```
Comment (mcwitt) [inline timemachine/cpp/src/math_utils.cuh:3]: Should we move `surreah.cuh` to `attic/` instead of deleting?

Comment (proteneer) [reply timemachine/cpp/src/math_utils.cuh:3]: I think that makes sense. It has some pedagogical utility (demonstrating how to implement forward mode AD via operator overloading).

File: timemachine/lib/potentials.py (hunk 1 above)
Comment (mcwitt) [reply timemachine/lib/potentials.py:140]: Fully deleted in 68bff6e

File: timemachine/cpp/src/math_utils.cuh (hunk 2 above)
Comment (mcwitt) [reply timemachine/cpp/src/math_utils.cuh:3]: Resurrected and moved to `attic/` in 4847d60

File: timemachine/cpp/src/kernels/k_nonbonded.cuh [hunk 3]
```diff
@@ -572,100 +480,50 @@ void __device__ v_nonbonded_unified(
...
 
-    RealType dq_dl_j = atom_j_idx < N ? dp_dl[atom_j_idx * 3 + 0] : 0;
-    RealType dsig_dl_j = atom_j_idx < N ? dp_dl[atom_j_idx * 3 + 1] : 0;
-    RealType deps_dl_j = atom_j_idx < N ? dp_dl[atom_j_idx * 3 + 2] : 0;
     RealType cw_j = atom_j_idx < N ? coords_w[atom_j_idx] : 0;
 
-    int is_vanilla =
-        (cw_i == 0 && dq_dl_i == 0 && dsig_dl_i == 0 && deps_dl_i == 0 && cw_j == 0 && dq_dl_j == 0 && dsig_dl_j == 0 &&
-         deps_dl_j == 0);
+    int is_vanilla = cw_i == 0 && cw_j == 0;
```
Comment (proteneer) [inline timemachine/cpp/src/kernels/k_nonbonded.cuh:518]: oh this is nice!

File: timemachine/cpp/src/nonbonded_all_pairs.hpp [hunk 4]
```diff
@@ -65,7 +62,7 @@ template <typename RealType, bool Interpolated> class NonbondedAllPairs : public
 
     void hilbert_sort(const double *d_x, const double *d_box, cudaStream_t stream);
 
-    std::array<k_nonbonded_fn, 16> kernel_ptrs_;
+    std::array<k_nonbonded_fn, 8> kernel_ptrs_;
```
Comment (proteneer) [inline timemachine/cpp/src/nonbonded_all_pairs.hpp:65]: nice! this should speed up compile times too.

File: tests/test_nonbonded.py [hunk 5]
```diff
@@ -281,47 +279,6 @@ def test_nblist_box_resize(self):
...
-
-            print(du_dx, du_dp, du_dl, u)
-
-            assert du_dl == 0.0
-
-    def test_fma_compiler_bug(self):
-
-        # this test case deals with a rather annoying fma compiler bug in CUDA.
-        # see https://github.com/proteneer/timemachine/issues/386
-        fp = gzip.open("tests/data/repro.pkl.gz", "rb")  # This assumes that primes.data is already packed with gzip
```
Comment (mcwitt) [inline tests/test_nonbonded.py:307]: This test relies on `NonbondedInterpolated` (via the pickle), and anyway the issue was seen in `du_dl`, which we don't compute anymore.

**Review by badisa — APPROVED** [review]
Looks great! The tests are faster and the benchmarks seem about 10ns faster even!

**Review by maxentile — APPROVED** [review]
Nice! This looks complete and correct as far as I can tell, and is a big simplification.

File: timemachine/cpp/src/nonbonded_interaction_group.cu [hunk 6]
```diff
@@ -264,12 +246,10 @@ void NonbondedInteractionGroup<RealType, Interpolated>::execute_device(
...
-    if (P != M * N_ * 3) {
+    if (P != N_ * 3) {
         throw std::runtime_error(
-            "NonbondedInteractionGroup::execute_device(): expected P == M*N_*3, got P=" + std::to_string(P) +
-            ", M*N_*3=" + std::to_string(M * N_ * 3));
+            "NonbondedInteractionGroup::execute_device(): expected P == N_*3, got P=" + std::to_string(P) +
+            ", N_*3=" + std::to_string(N_ * 3));
     }
 
     // identify which tiles contain interpolated parameters
```
[the remark this answers is not in the document]
Comment (mcwitt) [reply timemachine/cpp/src/nonbonded_interaction_group.cu]: Thanks, good catches. Fixed in be57eff.

File: timemachine/cpp/src/nonbonded_all_pairs.cu [hunk 7]
```diff
@@ -324,15 +301,8 @@ void NonbondedAllPairs<RealType, Interpolated>::execute_device(
     }
 
     // do parameter interpolation here
```
[the remark this answers is not in the document]
Comment (mcwitt) [reply timemachine/cpp/src/nonbonded_all_pairs.cu:301]: Removed in be57eff

**Pull request merged**